consensus/safety-rules/src/tests/suite.rs

```diff
@@ -600,3 +607,144 @@ fn test_sign_proposal_with_early_preferred_round(func: Callback) {
         Error::InvalidQuorumCertificate("Preferred round too early".into())
     );
 }
+
+fn test_uninitialized_signer(func: Callback) {
+    // Testing for an uninitialized Option<ValidatorSigner>
+
+    let (mut safety_rules, signer) = func();
+
+    let (proof, genesis_qc) = make_genesis(&signer);
+    let round = genesis_qc.certified_block().round();
+
+    let a1 = test_utils::make_proposal_with_qc(round + 1, genesis_qc, &signer);
+    let err = safety_rules.update(a1.block().quorum_cert()).unwrap_err();
+    assert_eq!(err, Error::NotInitialized);
+
+    safety_rules.initialize(&proof).unwrap();
+    assert!(safety_rules.update(a1.block().quorum_cert()).is_ok());
+}
+
+#[allow(dead_code)]
+fn test_missing_validator_key(func: Callback) {
+    // Tests for fetching a missing validator key from persistent storage.
+    // Ideally safety rules should panic
+
+    let (mut safety_rules, signer) = func();
+    let (proof, genesis_qc) = make_genesis(&signer);
+    let round = genesis_qc.certified_block().round();
+
+    safety_rules.initialize(&proof).unwrap();
+
+    let a1 = test_utils::make_proposal_with_qc(round + 1, genesis_qc, &signer);
+
+    let mut next_epoch_state = EpochState::empty();
+    next_epoch_state.epoch = 1;
+    let rand_signer = ValidatorSigner::random([0xfu8; 32]);
+    next_epoch_state.verifier =
+        ValidatorVerifier::new_single(signer.author(), rand_signer.public_key());
+    let a2 = test_utils::make_proposal_with_parent_and_overrides(
+        vec![],
+        round + 2,
+        &a1,
+        Some(&a1),
+        &signer,
+        Some(1),
+        Some(next_epoch_state),
+    );
+    assert!(safety_rules.update(a2.block().quorum_cert()).is_err());
+}
+
+fn test_missing_validator(func: Callback) {
```

Comment:
I think the test description could be cleaned up to be inclusive of what you're doing...

I think this test is too complicated for what it is attempting to do:
Ensure a proposal from a validator is within the ValidatorSet.

Then I would just get straight to the test that is... passing in a signed proposal from someone not in the validator set. No need to create a new verifier just a new signer and you're good to go.

Reply:
I've already done the test you mentioned in sign_proposal_with_bad_signer(). The difficulty of testing key reconciliation during epoch changes is two-fold: 1) to emulate epoch changes, because only in this way the code can be exercised, 2) to achieve high coverage, which requires me to carefully construct inputs in various forms -- each fails one stage of test but will fail progressively. Does this explain why my tests seem to more redundant than other tests? :)